# Исследование надежности заемщиков


<h2>  Описание проекта:  </h2> 
    <p>
<b>Заказчик</b> — кредитный отдел банка. <p> Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.<p>
     проверяемые гипотезы:
<ol type="1">
<li>Есть ли зависимость между количеством детей и возвратом кредита в срок?</li>
<li>Есть ли зависимость между уровнем дохода и возвратом кредита в срок? </li>
<li>Как разные цели кредита влияют на его возврат в срок? </li>
</ol> 

<b>Входные данные:</b> — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга.
    
<h2>Описание данных:</h2>
<p>
<ol>
<li>children — количество детей в семье</li>
<li>days_employed — общий трудовой стаж в днях</li>
<li>dob_years — возраст клиента в годах</li>
<li>education — уровень образования клиента</li>
<li>education_id — идентификатор уровня образования</li>
<li>family_status — семейное положение</li>
<li>family_status_id — идентификатор семейного положения</li>
<li>gender — пол клиента</li>
<li>income_type — тип занятости</li>
<li>debt — имел ли задолженность по возврату кредитов</li>
<li>total_income — ежемесячный доход</li>
<li>purpose — цель получения кредита  </li>

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



<div class="alert alert-info">
<b>Комментарий ревьера:</b> Проверка начинается с <a href=#link>шага 3</a></div>   

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for item in data['income_type'].unique():
    data.loc[(data['income_type'] == item) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == item), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for inc_type in data['income_type'].unique():
    data.loc[(data['income_type'] == inc_type) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == inc_type), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

<div id='link'></div> 

Представленный набор данных содержит <b>21524</b> записи статистики платежеспособности клиентов банка. <p>
В наборе присутствовали <b>2174</b> пропущенных  значения в столбцах  <b>days_employed (общий трудовой стаж в днях) и total_income (ежемесячный доход)</b> которые были заполнены медианными значениями в соответствии с категориями, к которым они относятся. <p>
    <ol>
<li> В столбце с информацией о числе детей клиентов (children) обнаружены и удалены строки с аномальными значениями.</li>
<li>Для простоты обработки и последующей категоризации данные о ежемесячном доходе были приведены к целочисленному типу (отброшена дробная часть).</li>
<li>Обработана и удалена 71 строка содержащая дубликаты данных.</li>
Проведена категоризация по:
<ol>
    <li>ежемесячному уровню дохода, добавлены категории A, B, C, D, E которые соответствуют доходу от 1000000 и выше, от 200000 до  1000000, от 50000 до  200000, от 30000 до 50000 и до 30000 соответственно</li>
   <li> целям на которые брался кредит (обобщены категории)</li> </ol></ol>

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Описание данных

- `'children'` — количество детей в семье
- `'days_employed'` — общий трудовой стаж в днях
- `'dob_years'` — возраст клиента в годах
- `'education'` — уровень образования клиента
- `'education_id'` — идентификатор уровня образования
- `'family_status'` — семейное положение
- `'family_status_id'` — идентификатор семейного положения
- `'gender'` — пол клиента
- `'income_type'` — тип занятости
- `'debt'` — имел ли задолженность по возврату кредитов
- `'total_income'` — ежемесячный доход
- `'purpose'` — цель получения кредита
- `total_income_category` — категории дохода
- `purpose_category` — категории целей кредита 
- `debt_category` — категории задолженностей 
- `gender_category` — категории пола 
- `family_status_category` — категории семейного положения
- `children_category` — категории по числу детей

In [22]:
data['gender'].unique()

array(['F', 'M', 'XNA'], dtype=object)

In [23]:
def categorize_children(count:int)->str:
    if count == 0:
        return 'нет детей'
    elif count == 1:
        return 'один ребенок'
    elif count == 2:
        return 'двое детей'
    elif count == 3:
        return 'трое детей'
    elif count == 4:
        return 'четверо детей'
    else:
        return 'пятеро детей'

функция для создания категорий

In [24]:
data['children_category'] = data['children'].apply(categorize_children)

добавляю читаемые данные о детях

In [25]:
data['debt_category'] = data['debt'].apply(lambda x: 'Имел задолженность' if x == 1 else 'Не имел задолженность')

добавляю категории по задолженностям

проверяю данные в наборе по половой принадлежности клиента, есть значение, которое нельзя сопоставить с полом человека. нужно проверить сколько записей соответствуют этому значению.

In [26]:
len(data.loc[data['gender'] == 'XNA'])

1

In [27]:
data = data.loc[(data['gender'] != 'XNA')]

данные можно удалить, так как это одна запись и существенного вклада в расчеты она внести не сможет

In [28]:
data['gender_category'] = data['gender'].apply(lambda x: 'Мужчина' if x =='M' else 'Женщина')

добавляю категории по половой принадлежности клиента

In [29]:
set(data.family_status)

{'Не женат / не замужем',
 'в разводе',
 'вдовец / вдова',
 'гражданский брак',
 'женат / замужем'}

добавим категории для семейного положния

In [30]:
data['family_status_category'] = \
data['family_status'].apply(lambda x: 'В браке' if x =='женат / замужем' else 'Не в браке')

In [31]:
def get_pivot_table(get_index:list, get_func:list=['sum', 'count', 'mean'], data_frame=data)->pd.DataFrame:

    data_pivot_table = data.pivot_table(
        index=get_index,
        values=['debt'],
        aggfunc=get_func)
    data_pivot_table = data_pivot_table.droplevel(1, axis=1)
    if 'mean' in data_pivot_table.columns:
        data_pivot_table.rename({'mean':'correlation'}, axis=1, inplace=True)
    return data_pivot_table

функция для создания сводной таблицы, по задолжнностям и с расчетом отношения общего числа задолженностей к общему числу записей для категории

In [32]:
my_pivot_table = get_pivot_table(['children','children_category'])

In [33]:
my_pivot_table

sum  count  correlation
children children_category                          
0        нет детей          1063  14090     0.075444
1        один ребенок        444   4808     0.092346
2        двое детей          194   2052     0.094542
3        трое детей           27    330     0.081818
4        четверо детей         4     41     0.097561
5        пятеро детей          0      9     0.000000

данных для анализа клиентов, имеющих пятерых детей, недостаточно. 
целесообразно исключить эту категорию из анализа текущей гипотезы.

In [34]:
my_pivot_table = my_pivot_table.reset_index().loc[
    my_pivot_table.reset_index()['children'] < 5]

my_pivot_table.sort_values(by=('correlation'))[
    ['children_category', 'sum', 'count', 'correlation']]

children_category   sum  count  correlation
0         нет детей  1063  14090     0.075444
3        трое детей    27    330     0.081818
1      один ребенок   444   4808     0.092346
2        двое детей   194   2052     0.094542
4     четверо детей     4     41     0.097561

зависимость в подученных данных прослеживается, так число просроченных кредитов у клиентов, которые не имеют детей меньше, чем у клиентов, которые имеют от одного ребенка и больше.
дополнительно нужно проверить лучшую группу для выяснения наиболее выгодного клиента и сопоставить данные выявленных клиентов на данных без учета детей

In [35]:
data['children_bool'] = data.loc[data['children'] < 5, 'children'].apply(lambda x: 'нет детей' if x == 0 else 'есть дети')

In [36]:
my_pivot_table = get_pivot_table(get_index=['children_bool'],
                                data_frame=data.loc[data['children_bool'] != None])

In [37]:
my_pivot_table.sort_values(by=('correlation'))

sum  count  correlation
children_bool                          
нет детей      1063  14090     0.075444
есть дети       669   7231     0.092518

In [38]:
len_data = len(data.loc[data['children'] < 5, 'children'])
for i in range(len(my_pivot_table)):
    print(f'кредиты клиентов у которых {my_pivot_table.index[i]} составляют {my_pivot_table["count"][i] / len_data:.2%} от общего числа обращений',
         f'и процент допущенных задолженностей по категории равен {my_pivot_table["correlation"][i]:.2%}')

кредиты клиентов у которых есть дети составляют 33.91% от общего числа обращений и процент допущенных задолженностей по категории равен 9.25%
кредиты клиентов у которых нет детей составляют 66.09% от общего числа обращений и процент допущенных задолженностей по категории равен 7.54%


данные подтверждяют промежуточные выводы, зависимость между количеством детей и возвратом кредита в срок - прослеживается. <p>
Кредиты клиентов у которых нет детей составляют `66.09%` от общего числа обращений и процент допущенных задолженностей по категории равен `7.54%` <p>
Кредиты клиентов у которых есть дети составляют `33.91%` от общего числа обращений и процент допущенных задолженностей по категории равен `9.25%` <p>

проверим лучшую группу 

In [39]:
my_pivot_table = get_pivot_table(['children', 'children_category', 'gender_category'])

In [40]:
my_pivot_table.reset_index()[
    my_pivot_table.reset_index()['children'] == 0].sort_values(by='correlation')

children children_category gender_category  sum  count  correlation
0         0         нет детей         Женщина  592   9520     0.062185
1         0         нет детей         Мужчина  471   4570     0.103063

женшина без детей - лучший заемщик. 
прооверим зависимость возвратов кредитов этой группы от семейного положения

In [41]:
my_pivot_table = get_pivot_table(['children', 'gender_category', 'family_status_category'])

In [42]:
my_pivot_table.reset_index()[
    (my_pivot_table.reset_index()['children'] == 0) &
    (my_pivot_table.reset_index()['gender_category'] == 'Женщина')
].sort_values(by='correlation')

children gender_category family_status_category  sum  count  correlation
0         0         Женщина                В браке  291   4885     0.059570
1         0         Женщина             Не в браке  301   4635     0.064941

In [43]:
my_pivot_table.reset_index()[
    (my_pivot_table.reset_index()['children'] == 0) &
    (my_pivot_table.reset_index()['gender_category'] == 'Мужчина')
].sort_values(by='correlation')

children gender_category family_status_category  sum  count  correlation
2         0         Мужчина                В браке  225   2583     0.087108
3         0         Мужчина             Не в браке  246   1987     0.123805

гипотеза о зависимости числа детей на возврат кредита подтвердилась, а также выяснилось, что лучшими заемщиками, допускающих наименьшее число задолженностей, стали женщины в браке.
нужно проверить эту гипотезу на полном наборе данных, для всех клиентов

In [44]:
get_pivot_table(['gender_category', 'family_status_category']).sort_values(by='correlation')

sum  count  correlation
gender_category family_status_category                         
Женщина         В браке                 526   7710     0.068223
                Не в браке              463   6382     0.072548
Мужчина         В браке                 401   4551     0.088113
                Не в браке              342   2687     0.127279

по общим данным видно, что женщины чаще берут кредиты и допускают меньше задолженностей по их возврату.

In [45]:
my_pivot_table = get_pivot_table(['gender_category', 'debt_category'], ['count'])
my_pivot_table

count
gender_category debt_category               
Женщина         Имел задолженность       989
                Не имел задолженность  13103
Мужчина         Имел задолженность       743
                Не имел задолженность   6495

In [46]:
list_pivot_table = list(my_pivot_table['count'])
print(f'{list_pivot_table[0] / list_pivot_table[1]:.2%} женщин допустило задолженность по кредиту')
print(f'{list_pivot_table[2] / list_pivot_table[3]:.2%} мужчин допустило задолженность по кредиту')

7.55% женщин допустило задолженность по кредиту
11.44% мужчин допустило задолженность по кредиту


**Гипотеза:** Есть ли зависимость между количеством детей и возвратом кредита в срок?

<div class="alert alert-info">
<p>
<b>Вывод:</b>  <p>Зависимость между количеством детей и возвратом кредита в срок - <b>прослеживается</b>.<p>

Кредиты клиентов у которых нет детей составляют 66.09% от общего числа обращений и процент допущенных задолженностей по категории равен 7.54%<p>

Кредиты клиентов у которых есть дети составляют 33.91% от общего числа обращений и процент допущенных задолженностей по категории равен 9.25%
</div>  

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [47]:
my_pivot_table = get_pivot_table(['family_status_category']).sort_values(by='correlation')
my_pivot_table

sum  count  correlation
family_status_category                         
В браке                 927  12261     0.075606
Не в браке              805   9069     0.088764

проверим данную гипотезу на более широкой категории

In [48]:
len_data = my_pivot_table['count'].sum()
for i in range(len(my_pivot_table)):
    print(f'кредиты клиентов {my_pivot_table.index[i]} составляют {my_pivot_table["count"][i] / len_data:.2%} от общего числа обращений',
         f'и процент допущенных задолженностей по категории равен {my_pivot_table["correlation"][i]:.2%}')

кредиты клиентов В браке составляют 57.48% от общего числа обращений и процент допущенных задолженностей по категории равен 7.56%
кредиты клиентов Не в браке составляют 42.52% от общего числа обращений и процент допущенных задолженностей по категории равен 8.88%


In [49]:
my_pivot_table = get_pivot_table(['family_status']).sort_values(by='correlation')

In [50]:
my_pivot_table

sum  count  correlation
family_status                                 
вдовец / вдова          63    951     0.066246
в разводе               84   1189     0.070648
женат / замужем        927  12261     0.075606
гражданский брак       385   4133     0.093153
Не женат / не замужем  273   2796     0.097639

In [51]:
len_data = len(data)
for item in range(len(my_pivot_table)):
    print(f'доля клиентов с симейным положением {my_pivot_table.index[item]}',
          f'составляет {my_pivot_table["count"][item] / len_data:.2%} от общего числа обращений',
          f'и процент допущенных задолженностей по категории равен {my_pivot_table["correlation"][item]:.2%}')

доля клиентов с симейным положением вдовец / вдова составляет 4.46% от общего числа обращений и процент допущенных задолженностей по категории равен 6.62%
доля клиентов с симейным положением в разводе составляет 5.57% от общего числа обращений и процент допущенных задолженностей по категории равен 7.06%
доля клиентов с симейным положением женат / замужем составляет 57.48% от общего числа обращений и процент допущенных задолженностей по категории равен 7.56%
доля клиентов с симейным положением гражданский брак составляет 19.38% от общего числа обращений и процент допущенных задолженностей по категории равен 9.32%
доля клиентов с симейным положением Не женат / не замужем составляет 13.11% от общего числа обращений и процент допущенных задолженностей по категории равен 9.76%


**Гипотеза:**  Есть ли зависимость между семейным положением и возвратом кредита в срок?

<div class="alert alert-info">
<p>
<b>Вывод:</b>  <p>Зависимость между семейным положением и возвратом кредита в срок - прослеживается.<p>
<p>
клиенты котрые состоят в браке или когда-либо состоявшие в браке допускают меньшее число задолженностей так<p>
доля клиентов с симейным положением вдовец / вдова составляет 4.46% от общего числа обращений и процент допущенных задолженностей по категории равен 6.62%<p>
доля клиентов с симейным положением в разводе составляет 5.57% от общего числа обращений и процент допущенных задолженностей по категории равен 7.06%<p>
доля клиентов с симейным положением женат / замужем составляет 57.48% от общего числа обращений и процент допущенных задолженностей по категории равен 7.56%<p>
доля клиентов с симейным положением гражданский брак составляет 19.38% от общего числа обращений и процент допущенных задолженностей по категории равен 9.32%<p>
доля клиентов с симейным положением Не женат / не замужем составляет 13.11% от общего числа обращений и процент допущенных задолженностей по категории равен 9.76%<p>
кредиты клиентов не состоящих в браке составляют 42.52% от общего числа обращений и процент допущенных задолженностей по категории равен 8.88%
</div>  

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [52]:
my_pivot_table = get_pivot_table(['total_income_category']).sort_values(by='correlation')
my_pivot_table

sum  count  correlation
total_income_category                          
D                        21    349     0.060172
B                       354   5013     0.070616
A                         2     25     0.080000
C                      1353  15921     0.084982
E                         2     22     0.090909

In [53]:
len_data = len(data)
for item in range(len(my_pivot_table)):
    print(f'доля клиентов с уровнем дохода {my_pivot_table.index[item]}',
          f'составляет {my_pivot_table["count"][item] / len_data:.2%} от общего числа обращений')

доля клиентов с уровнем дохода D составляет 1.64% от общего числа обращений
доля клиентов с уровнем дохода B составляет 23.50% от общего числа обращений
доля клиентов с уровнем дохода A составляет 0.12% от общего числа обращений
доля клиентов с уровнем дохода C составляет 74.64% от общего числа обращений
доля клиентов с уровнем дохода E составляет 0.10% от общего числа обращений


учитывая большую разницу в численности групп клиентов, распределенных по уровню дохода, расчетами для групп A, D, и E можно пренебречь 

In [54]:
my_pivot_table = my_pivot_table.reset_index()[
    (my_pivot_table.reset_index()['total_income_category'] == 'B') |
    (my_pivot_table.reset_index()['total_income_category'] == 'C')
].sort_values(by='correlation')
my_pivot_table

total_income_category   sum  count  correlation
1                     B   354   5013     0.070616
3                     C  1353  15921     0.084982

In [55]:
len_data = my_pivot_table['count'].sum()
#len_data = len(data)
for item in my_pivot_table.index:
    print(f'доля клиентов с уровнем дохода {my_pivot_table["total_income_category"][item]}',
          f'составляет {my_pivot_table["count"][item] / len_data:.2%} от общего числа обращений'
          f'и процент допущенных задолженностей по категории равен {my_pivot_table["correlation"][item]:.2%}')

доля клиентов с уровнем дохода B составляет 23.95% от общего числа обращенийи процент допущенных задолженностей по категории равен 7.06%
доля клиентов с уровнем дохода C составляет 76.05% от общего числа обращенийи процент допущенных задолженностей по категории равен 8.50%


**Гипотеза:** Есть ли зависимость между уровнем дохода и возвратом кредита в срок? 

 Вывод:

Прослеживается прямая зависимость между уровнем дохода и возвратом кредита в срок.

Так доля клиентов с уровнем дохода B составляет 23.95% от общего числа обращенийи процент допущенных задолженностей по категории равен 7.06%.

А доля клиентов с уровнем дохода C составляет 76.05% от общего числа обращенийи процент допущенных задолженностей по категории равен 8.50%.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [56]:
my_pivot_table = get_pivot_table(['purpose_category'])

In [57]:
my_pivot_table.sort_values(by='correlation', inplace=True)

In [58]:
my_pivot_table

sum  count  correlation
purpose_category                                 
операции с недвижимостью  780  10750     0.072558
проведение свадьбы        183   2313     0.079118
получение образования     369   3988     0.092528
операции с автомобилем    400   4279     0.093480

In [59]:
for i in range(len(my_pivot_table)):
    print(f'кредиты на {my_pivot_table.index[i]} составляют {my_pivot_table["count"][i] / len(data):.2%} от общего числа обращений',
         f'и процент допущенных задолженностей по категории равен {my_pivot_table["correlation"][i]:.2%}')

кредиты на операции с недвижимостью составляют 50.40% от общего числа обращений и процент допущенных задолженностей по категории равен 7.26%
кредиты на проведение свадьбы составляют 10.84% от общего числа обращений и процент допущенных задолженностей по категории равен 7.91%
кредиты на получение образования составляют 18.70% от общего числа обращений и процент допущенных задолженностей по категории равен 9.25%
кредиты на операции с автомобилем составляют 20.06% от общего числа обращений и процент допущенных задолженностей по категории равен 9.35%


**Гипотеза:** Как разные цели кредита влияют на его возврат в срок?

**Вывод:** 
- выдача кредитов на `операций с недвижимостью` составляют `50,40%` от общего числа всех займов, при этом имеют самый `низкий` процент (`7.26%`)  допущенных задолженностей.
- выдача кредитов на `проведение свадьбы` составляют `10.84%` от общего числа всех займов и имеют `средний` процент (`7.91%`) допущенных задолженностей .
- выдача кредитов на `операции с автомобилем` и `получение образования` составляют `20.06%` и `18.70%` соответственно, от общего числа всех займов, при этом имеют самый `худшие` показатели допущенных задолженностей `9.35%` и`9.25%` соответственно.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

**Пропуски могут появлятся:**
- из-за ошибок программного обеспечения при выгрузке или вводе данных;
- изменения структуры БД, при которой данные по категории ранее не собирались;
- некорректного ввода данных пользователем или сотрудником.

<div class="alert alert-info">
    если говорить касательно данного набора данных, можно предположить, что:
    <ol><li>пропуски данных в полях days_employed и total_income обусловлены отказом клиента предоставить данные для обработки, также возможно данные поля не обозначены как "обязательные к заполнению" и возможен вариант с более позднем добавлением столбцов в БД, но при этом поля с отсутствующими значениями находились бы в начале набора, однако возможна была произведена их сортировка по столбцам не попавшим в датасет.</li>
    <li>некорректных значений в полях children, отрицательных значений в полях total_income и некорректном значении поле gender это скорее всего ошибки оператора при вводе данных.</li>
    <li> выбросы days_employed для групп пенсионер и безработный вероятнее всего вызваны заполнением данных некорректными значениями при работе ПО (установка значения по умолчанию, для данных категорий) </li>
    <li> появление дубликатов это скорее всего погрешности вывода данных из базы</li></ol>
    </div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Заполнение пропусков медианным значением не окажет, или почти не окажет влияние на конечный расчет по данным  

### Шаг 4: общий вывод.

<ol>
    <li>Проверка гипотезы о зависимости между количеством детей и возвратом кредита в срок показала, следующее: </li>
    <ol>
        Кредиты клиентов у которых нет детей составляют <b>66.09%</b> от общего числа обращений и процент допущенных задолженностей по категории равен <b>7.54%</b> 
        Кредиты клиентов у которых есть дети составляют <b>33.91%</b> от общего числа обращений и процент допущенных задолженностей по категории равен <b>9.25%</b> </ol>
Из полученных результатов можно сделать вывод, что клиенты без детей, задолженности по кредитам допускают гораздо реже и чаще обращаются за заемными средствами.
    <li>Проверка гипотезы о зависимости между семейным положением и возвратом кредита в срок, дала следующее результаты:</li>
    <ol> Наилучший процент возврата кредита у вдовцов, число задолженностей у данной группы составляет <b>6.62%</b> численность группы - <b>4.46%</b> от общего числа обращений, у разведенных процент задолженностей немного выше и равен <b>7.06%</b> численность данной группы также немного выше - <b>5.57%</b>, клиенты состоящие в браке заняли третью позицию по числу плохих кредитов с процентом задолженностей <b>7.56%</b> при этом это самая многочисленная группа <b>57.48%</b> от общего числа клиентов. Аутсайдерами в данной группе стали клиенты состоящие в гражданском браке и не обремененные семьей, проценты по задолженностям составляют <b>9.32%, и 9.76%</b> при численности <b>19.38% и 13.11%</b> соответственно. 
При объединении клиентов не состоящих в браке в одну категорию ее размер составил <b>42.52%</b> и процент допущенных задолженностей получился в размере <b>8.88%</b>
    </ol>
После обработки результатов прослеживается тенденция к тому, что клиенты банка, официально оформившие свои отношения, более ответственно относятся к взятым на себя обязательствам в том числе и по отношению к заемным средствам.

<li>При анализе данных были получены результаты обратной зависимости уровня дохода от задолженностей по займам. </li>
    <ol>
        <li>Так доля клиентов с уровнем дохода <b>от 200000 до 1000000</b> составляет <b>23.95%</b> от общего числа обращений и процентом допущенных задолженностей по категории - <b>7.06%</b>.</li>
        <li>А доля клиентов с уровнем дохода <b>от 50000 до 200000</b> составляет <b>76.05%</b> от общего числа обращений и  процент допущенных задолженностей для данной группы - <b>8.50%</b>.</li>
    </ol>
То есть при уменьшении уровня дохода клиента вероятность возникновения у него задолженности увеличивается.
    
<li>Анализ целей кредитов и их влияния на возникновение задолженности:</li>
    <ol>
        <li>Выдача кредитов на операций с недвижимостью составляют <b>50,40%</b> от общего числа всех займов, при этом имеют самый низкий процент <b>(7.26%)</b> допущенных задолженностей.</li>
        <li>Выдача кредитов на проведение свадьбы составляют <b>10.84%</b> от общего числа всех займов и имеют средний процент <b>(7.91%)</b> допущенных задолженностей.</li>
        <li>Выдача кредитов на операции с автомобилем и получение образования составляют <b> 20.06% и 18.70% </b> соответственно, от общего числа всех займов, при этом имеют самый худшие показатели допущенных задолженностей <b>9.35% и 9.25% </b> соответственно.</li>
    </ol>

<li>Также при проверке первой гипотезы было обнаружено, что женщины не зависимо от их семейного положения реже мужчин допускают факты задолженности по взятым кредитам <b>7.55%</b> и <b>11.44%</b> соответственно.</li> 
    <ol>если рассмотреть данные и по семейному положению, то картина распределения будет выглядеть следующим образом:
        <li>Женщины, состоящие в браке обращались за заемными средствами <b>7710</b> раз, допустив <b>526</b> случаев задолженности, что соответствует <b>6,82%</b></li>
       <li>Женщины, не состоящие в браке обращались за заемными средствами <b>6382</b> раза, допустив <b>463</b> случаев задолженности, что соответствует <b>7,25%</b></li>
       <li>Мужчина, состоящие в браке обращались за заемными средствами <b>4551</b> раз, допустили <b>401</b> случаев задолженности, что соответствует <b>8,81%</b></li>
       <li>Мужчина, не состоящие в браке обращались за заемными средствами <b>2687</b> раз, допустили <b>342</b> случаев задолженности, что соответствует <b>12,72%</b></li>  </ol>
<li>рекомендации по работе с данными:</li>
 <ol>
    <li>обратить внимание на заполняемость данных в частности поля соответствующие days_employed и total_income для категорий "пенсионеры" и "безработные".</li>
    <li>также проверить возможность ввода в поле gender только двух категорий характеризующих пол клиента</li>
    <li>провести работу с операторским составом для корректного ввода данных</li></ol></ol>